server/src/message/mod.rs

```diff
@@ -0,0 +1,667 @@
+        x: f32,
+        y: f32,
+    },
+    /// **world_state** message, as defined by [Protocol spec](https://github.com/LoungeCPP/Tatsoryk/wiki/Protocol-spec)
+    ///
+    /// **world_state** — full update of the world, sent by the server to all connected clients periodically (interval up to the implementation)
+    /// - `player_count` (u32) — count of all connected players
+    /// - `alive_players` (Player[]) — an array of all currently alive players, each containing:
+    ///   - `id` (u32) — ID of the player
+    ///   - `x` (f32) — current position X of the player
+    ///   - `y` (f32) — current position Y of the player
+    ///   - `move_x` (Optional&lt;f32&gt;) — current movement vector X of the player, if player is moving
+    ///   - `move_y` (Optional&lt;f32&gt;) — current movement vector Y of the player, if player is moving
+    /// - `alive_bullets` (Bullet[]) — an array of all currently alive bullets, each containing:
+    ///   - `id` (u32) — ID of the bullet
+    ///   - `x` (f32) — current position X of the bullet
+    ///   - `y` (f32) — current position Y of the bullet
+    ///   - `move_x` (f32) — current movement vector X of the bullet
+    ///   - `move_y` (f32) — current movement direction vector Y of the bullet
+    ///                      (movement direction vectors MUST be normalised, i.e. their magnitude MUST be equal to 1)
+    WorldState,
+    /// **start_moving** message, as defined by [Protocol spec](https://github.com/LoungeCPP/Tatsoryk/wiki/Protocol-spec)
+    ///
+    /// **start_moving** — sent by the client to the server when the player wants to start moving or change its movement direction
+    ///                    (i.e. presses/releases one or more movement keys, as long as at least one of them is still held)
+    /// - `move_x` (f32) — player's movement vector X
+    /// - `move_y` (f32) — player's movement vector Y
+    /// (movement vector SHOULD be normalised, but the server MUST NOT assume that it is)
+    StartMoving {
+        move_x: f32,
+        move_y: f32,
+    },
+    /// **stop_moving** message, as defined by [Protocol spec](https://github.com/LoungeCPP/Tatsoryk/wiki/Protocol-spec)
+    ///
+    /// **stop_moving** — sent by the client to the server when the player wants to stop moving (i.e. releases held movement keys)
+    StopMoving,
+    /// **fire** message, as defined by [Protocol spec](https://github.com/LoungeCPP/Tatsoryk/wiki/Protocol-spec)
+    ///
+    /// **fire** — sent by the client to the server when the player wants to fire (i.e. presses the mouse button)
+    /// - `move_x` (f32) — player's aiming vector X
+    /// - `move_y` (f32) — player's aiming direction vector Y (aiming direction vector SHOULD be normalised, but the server MUST NOT assume that it is)
+    Fire {
+        move_x: f32,
+        move_y: f32,
+    },
+}
+
+impl ToString for Message {
+    fn to_string(&self) -> String {
+        let mut values = BTreeMap::new();
+        let msg_type = match self {
```

Comment:
Is this really the best structure? Having to come here for every change to messages is going to. Can't each individual message cleanly implement it's own 'to_string'?